core/src/banking_stage.rs

```diff
@@ -701,27 +615,124 @@ impl BankingStage {
 
                         reached_end_of_slot = Some(EndOfSlot {
                             next_slot_leader: poh_recorder_locked.next_slot_leader(),
-                            working_bank: None,
+                            working_bank: Some(working_bank),
                         });
                         poh_recorder_lock_time
                     };
+
                     slot_metrics_tracker.increment_consume_buffered_packets_poh_recorder_lock_us(
                         poh_recorder_lock_time.as_us(),
                     );
+                }
 
-                    // `original_unprocessed_indexes` must have remaining packets to process
-                    // if not yet processed.
-                    assert!(!original_unprocessed_indexes.is_empty());
-                    true
+                // The difference between all transactions passed to execution and the ones that
+                // are retryable were the ones that were either:
+                // 1) Committed into the block
+                // 2) Dropped without being committed because they had some fatal error (too old,
+                // duplicate signature, etc.)
+                //
+                // Note: This assumes that every packet deserializes into one transaction!
+                consumed_buffered_packets_count +=
+                    packets_to_process_len.saturating_sub(retryable_transaction_indexes.len());
+
+                // Out of the buffered packets just retried, collect any still unprocessed
+                // transactions in this batch for forwarding
+                rebuffered_packet_count += retryable_transaction_indexes.len();
+                /*let has_more_unprocessed_transactions = deserialized_packet_batch
+                // TODO: no longer need to track batches, can just hand over to threads directly
+                .update_buffered_packets_with_new_unprocessed(
+                    &original_unprocessed_indexes,
+                    &retryable_transaction_indexes,
+                );*/
+
+                // TODO: Add all retryable transactions into `retryable_packets`.
+                for retryable_index in retryable_transaction_indexes {
+                    retryable_packets.push(packets_to_process[retryable_index].clone());
```

Comment:
wondering how much the repeat of clone/push would cost. This [bench test](https://github.com/solana-labs/solana/issues/24646#issuecomment-1108787440) sets all 128 transactions in same batch to write to same account, so there will be a lot of `account-in-use` retries. The result compare to `master` is not great.

Reply:
yeah it's definitely not ideal, but the retryable packets need to be stored somewhere to be reattempted later. If we had a global scheduler it might be able to minimize these, but open to suggestions

Reply:
The first Packet.clone from packet_batch is one time per packet, so a static overhead. But `clone` here to handle retryablies are dynamic, same packet may be cloned multiple times during runtime, so it's accumulated cost could be high. I'm wondering if it is operating on `Rc` instead of `Packet`, something like I did [here](https://github.com/solana-labs/solana/pull/24187#discussion_r858016836), would that be an alternative?

Reply:
It's operating on `DeserializedPacket`, not `Packet`, which looks like

```
pub struct DeserializedPacket {
    immutable_section: Rc<ImmutableDeserializedPacket>,
    pub forwarded: bool,
}
```

so most of it is already in an `Rc`

Reply:
right, it's just `Rc`